synapse/handlers/search.py

```diff
@@ -65,8 +67,17 @@ def get_old_rooms_from_upgraded_room(self, room_id):
             if not predecessor:
                 break
 
+            predecessor_room_id = predecessor["room_id"]
+
+            # Don't bother trying to search in rooms we're not in
+            in_room = yield self.auth.check_host_in_room(
```

Comment:
check_host_in_room is *expensive*. Also, I don't think it really does what you want: it checks if any of the users on the server are *currently* in the room, whereas for search it is more appropriate to consider rooms that the server has ever participated in.

we can actually tell this because `get_room_predecessor` will raise a `NotFoundError`. I'd suggest shuffling the loop around so that you don't add the room_id to `historical_room_ids` until you've called `get_room_predecessor`, and bailing out when you get a `NotFoundError`.